Extended subscription/unsubscription (selectable unsubscription process)

bures opened the pull request:
This PR represents some extensions and refactoring of the subscription/unsubscription process.

It addresses issues #218, #219, #220, #221, #222

It adds the re-subscribe feature, which was probably intended, but somehow not completed before. This means that when "Subscribe" link is clicked in unsubscribe confirmation, the form fields are pre-filled with the original values. The subscription mechanism (already existing) ensures that the subscription gets the same subscription cid as before.

Further, it contains some refactoring and code de-duplication to make future extensions of the subscription/unsubscription process easier. The most important change are:
- Sending emails moved solely to router. (Before it was partially in model and partially in router). Further, all subscription-related emails have been factored into helper methods, all in subscription-mail-helpers.js
- Confirmations functionality refactored to be generic (not specific to "subscribe"). It is now used to "subscribe", "change-address", and "unsubscribe". It can be further used for other actions that need email confirmation (even beyond the subscription/unsubscription process). For this reason, the confirmations have been moved to a dedicated model package.
- Some of the custom forms have been renamed to have more unified naming. Some more have been added to address #220, #219.

### Potential upgrade issues:
- The schema of the "data" field in the confirmation table differs to the previous version. It would be difficult to reconstruct it in the upgrade SQL-script. Thus, I decided to prune the contents of the confirmation table on schema upgrade. This means that all subscribe confirmations become invalid. The system will then give a message saying that the request is invalid and that subscription should be done again. Hopefully this solves the problem (though it will be annoying in some singular cases.)

- The PR adds a few more custom forms. If a list has some custom forms, the content of the new custom forms will be empty. It is thus necessary that after the upgrade, one goes through all existing custom forms and updates them according the the provided defaults.

### Testing:
I hope I tested the whole subscribe/manage/unsusbscribe features. I also tested the API endpoints, subscription widget and subscribe/unsubscribe in list management. However, it is possible that I forgot about some dependency or just didn't know about it. Please give it some try yourselves.

Thanks a lot. So, user management now? :-)
T.

witzig commented:
Awesome, thank you.

Concerning upgrade issues, I think custom forms fall back gracefully and thus this is only an issue for people who have their forms translated. Styling should be unaffected. 

> The schema of the "data" field in the confirmation table differs to the previous version. [...] Thus, I decided to prune the contents of the confirmation table on schema upgrade.

Shouldn't we keep those for legal reasons?

> I also scraped the "auto" parameter which automatically submits the unsubscription form when the link is clicked in a campaign email.

Regarding the List-Unsubscribe Header, would it be advisable to adhere to the [guidelines of the Certified Senders Alliance](https://certified-senders.eu/wp-content/uploads/CSA_list_unsubscribe.pdf)?

@andris9, I tested it too. Works great!

bures commented:
Thank you Roger for testing it.

>> The schema of the "data" field in the confirmation table differs to the previous version. [...] Thus, I decided to prune the contents of the confirmation table on schema upgrade.

> Shouldn't we keep those for legal reasons?

As far as I remember, the confirmations contained only pending confirmations. Once a confirmation was process, it was deleted. I preserved this functionality. Generally, the confirmations table should be rather empty and not much useful for any auditing / legal purposes.

>> I also scraped the "auto" parameter which automatically submits the unsubscription form when the link is clicked in a campaign email.

> Regarding the List-Unsubscribe Header, would it be advisable to adhere to the guidelines of the Certified Senders Alliance?

OK. Now I understand why it was there. :-) I'll reintroduce it and let you know.

witzig commented:
Oh, very nice! And sorry for the extra work. Not sure I should have pushed the draft in retrospect. :-)

bures commented:
No worries Roger. I'm very glad that you did the e2e tests. We needed to do them and this was an excellent opportunity. I also liked the way you did it. I followed your structure, just extended it a bit.

In particular, I wanted something that from the perspective of Mocha behaves as a single test, but itself is structured to steps. The abstraction of a use-case was very natural here because it describes interaction between a user and the system.

The idea here is that a Mocha test (being primarily a unit test) can generally be executed independently of other tests (provided that before and after hooks are called). From this perspective, the use-case is a single test, because one cannot just repeat the step that failed previously - instead all steps leading to it have to be repeated.

To support this better, I created a custom reporter for Mocha, which is aware of use-case steps and can display them. If another reporter is used, it interprets the whole use-case as one test and just the steps won't be displayed.

I now just would like to finish the use-case tests for all the different variants of subscription.
Once this is done, it can hopefully motivate us to create e2e tests for the new functionality.

witzig commented:
Oh sure, the extensions and clean up make a ton of sense. I also like your composition approach, quite elegant.

> I now just would like to finish the use-case tests for all the different variants of subscription.

Happy to help with this ...

witzig commented on test/e2e/tests/subscription-uc.js:
```diff
@@ -0,0 +1,106 @@
+        });
+
+        await step('System show a notice that confirms unsubscription', async () => {
+            await page.webUnsubscribedNotice.waitUntilVisible();
```
This works fine with `phantomjs` yet fails when using `firefox` as the `elementToWaitFor` (homepageButton) is already present in `mailSubscriptionConfirmed`. Sleeping a second helps, obviously, but perhaps a better approach might be to something like `textToWaitFor`?

bures replied:
Roger, thanks a lot for doing this interim testing. Really greatly appreciated. I suspected this could be an issue.

Well, one "correct" solution that comes to my mind would be to extend waitUntilVisible with a piece of custom browser code something like document.mailTrainRefreshAcknowledged=true.

Then we could have a new function waitUntilVisibleAfterRefresh which would first wait for the document to lose the mailTraiRefreshAcknowledged property. Only then it would start waiting for the element.

Another options is to be able to wait for multiple items on the page and ensure that always at least one is distinguishing.

What do you think? I would possibly do both since they address slightly different cases.

witzig replied:
In hindsight, I'm not sure whether using elements for the catch all `waitUntilVisible` was the right decision, at least for the subscription process. Perhaps it would make more sense to [wait for](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/lib/until.html) unique text, e.g. "Subscription Confirmed". This way we'd also get an essential text assertion for free and the links/buttons are tested no later than we're trying to click them.

> What do you think? I would possibly do both since they address slightly different cases.

Sounds good. `waitUntilVisibleAfterRefresh ` would solve this problem nicely. Waiting for multiple items makes sense too, yet (for this particular issue) there aren't many readily identifiable elements in the HTML generated by MJML.

bures commented:
@witzig - There a few more tests to that I want to do. Also I would like to fix the issue with the "auto" parameter.
However, this is what I believe could be more or less the way tests could look like.

I added the two things we discussed above. Fixed the reporting a bit. 

I introduced functions that save the current page (as HTML, url, and PNG). This can be invoked via API. It also happens automatically when a test fails - it gets stored in last-failed-e2e-test.xxx. This is quite useful when creating and debugging the tests.

I also extended the API of "useCase". It can be now called with .only, .skip - the same was as "it"/"test" can be marked with .skip/.only.

witzig commented:
Tomas, this is just fantastic. I'm greatly amused watching Firefox executing all these tests. :-)

> There a few more tests to that I want to do.

You've already dedicated quite some time towards the e2e tests, so I'd be happy to help with the remaining use cases you described. It's mostly assiduity work anyway and I think the foundation is laid and quite complete.

Regarding the List Unsubscribe Header, it's only been implemented recently with 745bfb0. Also here's a more detailed explanation [how it works with gmail](http://www.zettasphere.com/gmail-email-unsubscribe-header/). Given the example with the emails from Google Analytics, it looks like the auto parameter isn't too important after all ...

bures commented:
Perfect. Thanks. In that case, I'll only  implement the "auto" parameter in the afternoon and kindly ask you for help with the remaining tests.

How shall we do it. Merge this PR first once the "auto" parameter is in?
Or move it to a branch and finish it there? Or I can give you developer access to my fork (bures/mailtrain). As you like. Thanks.

witzig commented:
We both tested the remaining use cases already manually. So, yes, let's merge this once the "auto" parameter is in.

bures commented:
I'm a bit puzzled now. The List-Unsubscribe header has been introduced by 745bfb0, but it has later been deleted in 179e0d5ab39bfede012849e9a1a49f4e27802a39. 

@andris9 - it seems that the deletion was by you. Was it intentional?

bures commented:
I've added handling of the "auto" parameter to the unsubscription. However, the List-Unsubscribe header is currently not included in the emails (see my comment above).

andris9 commented:
@bures I think the List-Unsubscribe was removed because of custom unsubscribe links. Some users had their own custom unsubscribe pages and updated lists via API so they did not want the link to be there. Now I think that removing the link was a mistake, instead it should have been configurable option or something like that. I'm in favor putting the link back to the headers as default.

witzig commented:
I'm a bit puzzled too, as the List-Unsubscribe Header is still present. I reckon it's [generated by Nodemailer](https://community.nodemailer.com/2-0-0-beta/list-headers/) based on [this](https://github.com/Mailtrain-org/mailtrain/blob/684ba69d467e7167a09a95497e2ff8b386c48e6a/services/sender.js#L420-L422). @andris9, might this have been the reason for the removal in 179e0d5?

witzig commented:
@bures, could you add your git email to github, so your commits get properly linked and attributed? Thanks.

bures commented:
@witzig, good point. Can you please check now. I added the address to github. Thanks.

witzig commented:
@bures, your commits are now properly linked. Thanks.

witzig commented:
I pushed the remaining tests with 19a563b. Any other scenarios we should check?

Thoughts about integration with https://travis-ci.org?

bures commented:
@witzig - Thanks a lot for the remaining tests. I've fixed a few eslint problems and I'm commiting it now. 
As for the switchToList, I updated it such that the page object is constructed for every use-case. I believe this is cleaner as they are now self-contained.

Travis may be be good. Also it would make sense to automate the execution of eslint. I admit, I forgot about it when doing the last PR.

witzig commented:
@bures, no worries regarding eslint. I'm guilty myself about forgetting it, Andris can tell. :-)

Also, I should have told you that I already fixed the errors. I assumed merging master would be your first move before continuing work on the e2e stuff. Sorry about that. Fixing the conflicts in 731226d surely wasn't fun. :/

> Also it would make sense to automate the execution of eslint.

#242 would take care of that. And perhaps a pre-commit hook could be an option, too.

bures commented:
>  I assumed merging master would be your first move before continuing work on the e2e stuff.
Haha. That would be of course very natural thing to do if I had not forgotten about your commit (which I of course saw the day before). :-D

andris9 commented:
@witzig oh, yes, it seems that I moved the List header generation from one place to another and did not just remove it. Already forget about it. I think I re-implemented [List- header handling](https://nodemailer.com/message/list-headers/) for Nodemailer and wanted to test it with Mailtrain.
